src/plans/constrained_plan.jl

```diff
@@ -35,824 +12,409 @@ Describes the constrained objective
     )
 end
-#
-# Neither equality nor inequality yields an error
-#
 function ConstrainedManifoldObjective(
-    ::TF, ::TGF, ::Nothing, ::Nothing, ::Nothing, ::Nothing; kwargs...
-) where {TF,TGF}
-    return error(
+    objective::MO;
+    equality_constraints::EMO=nothing,
+    inequality_constraints::IMO=nothing,
+    kwargs...,
+) where {E<:AbstractEvaluationType,MO<:AbstractManifoldObjective{E},IMO,EMO}
+    if isnothing(equality_constraints) && isnothing(inequality_constraints)
+        @warn """
+        Neither the inequality and the equality constraints are provided.
+        Consider calling `get_objective()` on this constraint objective
+        and only work on the unconstrained objective instead.
         """
-  Neither inequality constraints `g`, `grad_g` nor equality constraints `h`, `grad_h` provided.
-  If you have an unconstraint problem, maybe consider using a `ManifoldGradientObjective` instead.
-  """,
+    end
+    return ConstrainedManifoldObjective{E,MO,IMO,EMO}(
+        objective, equality_constraints, inequality_constraints
     )
 end
-function ConstrainedManifoldObjective(
-    f::TF,
-    grad_f::TGF;
-    g=nothing,
-    grad_g=nothing,
-    h=nothing,
-    grad_h=nothing,
-    evaluation::AbstractEvaluationType=AllocatingEvaluation(),
-) where {TF,TGF}
-    return ConstrainedManifoldObjective(
-        f, grad_f, g, grad_g, h, grad_h; evaluation=evaluation
+
+@doc raw"""
+    ConstrainedProblem{
+        TM <: AbstractManifold,
+        O <: AbstractManifoldObjective
+        GR <: AbstractManifold
+        HR <: AbstractManifold
+    } <: AbstractManoptProblem{TM}
+
+A constrained problem might feature different ranges for the
+(vectors of) gradients of the equality and inequality constraints.
+
+The ranges are required in a few places to allocate memory and access elements
+correctly, they work as follows:
+
+Assume the objective is
+```math
+\begin{aligned}
+ \operatorname*{arg\,min}_{p ∈\mathcal{M}} & f(p)\\
+ \text{subject to } &g_i(p)\leq0 \quad \text{ for all } i=1,…,m,\\
+ \quad &h_j(p)=0 \quad \text{ for all } j=1,…,n.
+\end{aligned}
+```
+
+then the gradients can (classically) be considered as vectors of the
+components gradients, for example
+``\bigl(\operatorname{grad} g_1(p), \operatorname{grad} g_2(p), …, \operatorname{grad} g_m(p) \bigr)``.
+
+In another interpretation, this can be considered a point on the tangent space
+at ``P = (p,…,p) \in \mathcal M^m``, so in the tangent space to the [`PowerManifold`](@extref) ``\mathcal M^m``.
+The case where this is a [`NestedPowerRepresentation`](@extref) this agrees with the
+interpretation from before, but on power manifolds, more efficient representations exist.
+
+To then access the elements, the range has to be specified. That is what this
+problem is for.
+"""
+struct ConstrainedManoptProblem{
+    TM<:AbstractManifold,
+    O<:AbstractManifoldObjective,
+    GR<:Union{AbstractPowerRepresentation,Nothing},
```

Comment:
Why do you need to allow `Nothing` here?

Reply:
There is two reasons.
The main one is: There might just to be any inequality constraints `g`(oh I mixed up the two below).
The other one could probably be removed by proper constructors, when the user does not specify it we choose the default.

Reply:
Could you give an example where no `AbstractPowerRepresentation` works? I don't see it.

Reply:
Oh, they would always work. Just to be backward-compatible, I am not yet sure whether I always have one available I can fill in.
And if G does not exist I think it is just fair to have nothing here as well

Reply:
`NestedPowerRepresentation` should work fine I think as a default? Then we wouldn't have issues with `nothing` being there when it shouldn't.

Reply:
And if `g` does not exist itself the type would still exist? To me that feels a bit inconsistent – though mathematically not wrong.

Reply:
I see, using `Nothing` when there is no `g` makes sense, but only in that case.

Reply:
I think you are right, the fallback case to be compatible should be the Nested one.

Reply:
A few lines below you can see by default it is filled even with Nested